proposals/inline-classes.md

```diff
@@ -0,0 +1,163 @@
+- Inline class cannot have inner classes
+- Inline class must be a toplevel class 
+
+### Other restrictions
+
+The following restrictions are related to the usages of inline classes:
+
+- Referential equality (`===`) is prohibited for inline classes
+- vararg of inline class type is prohibited
+```
+inline class Foo(val s: String)
+
+fun test(vararg foos: Foo) { ... } // should be an error  
+```
+
+## Java interoperability  
+
+Each inline class has its own wrapper, which is boxed/unboxed by the same rules as for primitive types.
+Basically, rule for boxing can be formulated as follows: inline class is boxed when it is used as another type.
+Unboxed inline class is used when value is statically known to be inline class.
+
+Examples:
+```
+interface I
+
+inline class Foo(val i: Int) : I
+
+fun asInline(f: Foo) {}
+fun <T> asGeneric(x: T) {}
+fun asInterface(i: I) {}
+fun asNullable(i: Foo?) {}
+
+fun <T> id(x: T): T = x
+
+fun test(f: Foo) {
+    asInline(f)
+    asGeneric(f) // boxing
+    asInterface(f) // boxing
+    asNullable(f) // boxing
+    
+    val c = id(f) // boxing/unboxing, c is unboxed
+}
+```
+
+Since boxing doesn't have side effects as is, it's possible to reuse various optimizations that are done for primitive types.
+
+### Type mapping on JVM
+
+Depending on the underlying type of inline class and its declaration site, inline class type can be mapped to the underlying type or to the 
+wrapper type.  
+
+
+| Underlying Type \ Declaration Site | Not-null | Nullable | Generic |
+| --------------------------------------------- | --------------- | ---------------- | ---------------- |
+| **Reference type** | Underlying type | Underlying type | Wrapper type |
+| **Nullable reference type** | Underlying type  | Wrapper type | Wrapper type |
+| **Primitive type** | Underlying type  | Wrapper type | Wrapper type |
+| **Nullable primitive type** | Underlying type  | Wrapper type | Wrapper type |
+
+For example, from this table it follows that nullable inline class which is based on reference type is mapped to the underlying type:
+```
+inline class Name(val s: String)
+
+fun foo(n: Name?) { ... } // `Name?` here is mapped to String
+```
+
+## Methods from `kotlin.Any`
+
+Inline classes are indirectly inherited from `Any`, i.e. they can be assigned to a value of type `Any`, but only through boxing.
+
+Methods from `Any` (`toString`, `hashCode`, `equals`) can be useful for a user-defined inline classes and therefore should be customizable. 
+Methods `toString` and `hashCode` can be overridden as usual methods from `Any`. For method `equals` we're going to introduce new operator 
+that represents "typed" `equals` to avoid boxing for inline classes:
+```
+inline class Foo(val s: String) {
+    operator fun equals(other: Foo): Boolean { ... }
+}
+```
+
+Compiler will generate original `equals` method that is delegated to the typed version.  
```

Comment:
I consider harmful expose the private value for the `UInt` example, or the `component1` must have the same visibility of the `value` attribute

Reply:
Could you please elaborate? Do you have a use case? Maybe it's a point to allow inline data classes

Reply:
Oh, you're right, auto `component1` is useless.
Vice versa, it may be used by hand, for example:
```
inline class IntPair(private val value: Long) {
    operator fun component1() = value ushr 32
    operator fun component2() = value and 0xFFFFFFFF
}
```